### PR TITLE
As a user, I can upload CSV file

### LINKED ISSUE #7: As a user, I can upload csv
### Acceptance Criteria
- Implement upload CSV feature
  - UI to display uploaded keywords
  - Backend to process upload task
- Only authenticated user can use this feature
- Allow only CSV file type
- Allow only 1-1000 keywords CSV file
- Include unit test

### PR DESCRIPTION
**Pull request by gutakk**
## What happened
- [x] Create keyword upload page

## Insight
- [x] Add current user middleware
- [x] Add keyword upload handler with unit test
  - Validate file type
  - Validate CSV length
  - Save keywords to database
- [x] Refactor test

## Proof Of Work
- Keyword upload page
Screen Shot 2563-12-02 at 16 28 41

- Access to keyword upload page without login
Screen Shot 2563-12-02 at 16 30 01

- File uploaded successfully
Screen Shot 2563-12-02 at 16 29 21

```
 id  |          created_at           |          updated_at           | deleted_at |      keyword      | user_id 
-----+-------------------------------+-------------------------------+------------+-------------------+---------
 132 | 2020-12-02 09:31:11.351521+00 | 2020-12-02 09:31:11.351521+00 |            | Eden Hazard       |       8
 133 | 2020-12-02 09:31:11.351521+00 | 2020-12-02 09:31:11.351521+00 |            | Lionel Messi      |       8
 134 | 2020-12-02 09:31:11.351521+00 | 2020-12-02 09:31:11.351521+00 |            | Cristiano Ronaldo |       8
 135 | 2020-12-02 09:31:11.351521+00 | 2020-12-02 09:31:11.351521+00 |            | Kylian Mbappe     |       8
 136 | 2020-12-02 09:31:11.351521+00 | 2020-12-02 09:31:11.351521+00 |            | Neymar            |       8
 137 | 2020-12-02 09:31:11.351521+00 | 2020-12-02 09:31:11.351521+00 |            | Kevin De Bruyne   |       8
(6 rows)
```

- File uploaded unsuccessfully
Screen Shot 2563-12-02 at 16 29 06

### CODE REVIEW

File: controllers/keyword.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	engine.GET("/keyword", k.displayKeyword)
+	engine.POST("/keyword", k.uploadKeyword)
+}
+
+func (k *KeywordController) displayKeyword(c *gin.Context) {
+	html.RenderWithFlash(c, http.StatusOK, keywordView, keywordTitle, nil)
+}
+
+func (k *KeywordController) uploadKeyword(c *gin.Context) {
+	file, _ := c.FormFile("file")
```
Comment (olivierobert) [inline controllers/keyword.go]: How about using a struct like in the registration and login? Having a struct to validate params is great even if there is only one param 💪

File: controllers/keyword.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	// Validate if file is CSV type
+	if err := models.ValidateFileType(file.Header["Content-Type"][0]); err != nil {
+		html.RenderWithError(c, http.StatusBadRequest, keywordView, keywordTitle, err, nil)
+	}
+
+	filename := "dist/" + filepath.Base(file.Filename)
+	_ = c.SaveUploadedFile(file, filename)
+	csvfile, _ := os.Open(filename)
+	r := csv.NewReader(csvfile)
+	record, _ := r.ReadAll()
```
Comment (olivierobert) [inline controllers/keyword.go]: Line 38 -> 42 should also be moved outside of the controller for the same reasons I gave you before: it's the business domain 💍 There needs to be an API to perform CSV upload so the same logic would be needed. Encapsulating it into a new object would make it reusable e.g. `keywordFile` and a method `keywordFile.upload()`.

Great to move `ValidateCSVLength` and `SaveKeywords` to the keyword model 💪

File: models/keyword.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		return errors.New(fileLengthError)
+	}
+	return nil
+}
+
+func SaveKeywords(record [][]string) ([]Keyword, error) {
+	var keywords = []Keyword{}
+
+	// Check if record more than 0
+	if len(record) <= 0 {
```
Comment (olivierobert) [inline models/keyword.go]: Can this be really negative?

File: models/keyword.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+func ValidateCSVLength(row int) error {
+	if row <= 0 || row > 1000 {
+		return errors.New(fileLengthError)
+	}
+	return nil
+}
+
+func SaveKeywords(record [][]string) ([]Keyword, error) {
```
Comment (olivierobert) [inline models/keyword.go]: What will be the CSV format be like? I am not sure why it needs to be an array of array of strings 🤔

File: models/keyword_test.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+func TestValidateCSVLengthWithZeroRow(t *testing.T) {
+	result := ValidateCSVLength(0)
+
+	assert.Equal(t, "CSV file must contain between 1 to 1000 keywords", result.Error())
+}
+
+func TestValidateCSVLengthWithMinusOneRow(t *testing.T) {
+	result := ValidateCSVLength(-1)
```
Comment (olivierobert) [inline models/keyword_test.go]: How can this case even happen 🤔 ?

File: models/keyword_test.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	assert.Equal(t, "File must be CSV format", result.Error())
+}
+
+func TestValidateCSVLengthWithOneRow(t *testing.T) {
+	result := ValidateCSVLength(1)
+
+	assert.Equal(t, nil, result)
+}
+
+func TestValidateCSVLengthWithOneThoudsandRow(t *testing.T) {
```
Comment (olivierobert) [inline models/keyword_test.go]: The min/max values might change so it would better to named some tests in a more generic way e.g. `TestValidateCSVLengthWithMaxRowAllowed`

File: models/keyword_test.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	suite.Run(t, new(KeywordDBTestSuite))
+}
+
+func (s *KeywordDBTestSuite) TestSaveKeywordsWithValidParams() {
+	keywords := [][]string{
+		[]string{"Hazard"},
+		[]string{"Ronaldo"},
+		[]string{"Neymar"},
+		[]string{"Messi"},
+		[]string{"Mbappe"},
```
Comment (olivierobert) [inline models/keyword_test.go]: Based on my previous comments, wouldn't be simpler not to have the nested arrays?

```go
[]string{"Hazard", "Ronaldo", "Neymar", "Messi", "Mbappe"}
```

In the CSV parsing, the data could be cleaned to have this format, no?

File: models/setup_test.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package models
+
+import (
+	"os"
+	"testing"
+
+	"github.com/gin-gonic/gin"
+)
+
+func TestMain(m *testing.M) {
```
Comment (olivierobert) [inline models/setup_test.go:10]: Why adding this file now?

**Review by olivierobert — COMMENTED** [review]
Some early feedback 👀

File: models/keyword.go (hunk 4 above)
Comment (gutakk) [reply models/keyword.go]: As I used `ReadAll()` function and the signature look like this
```go
func (r *Reader) ReadAll() (records [][]string, err error) {...}
```

File: models/setup_test.go (hunk 8 above)
Comment (gutakk) [reply models/setup_test.go:10]: Because `TestMain` can declare only once per package [(REF)](https://medium.com/goingogo/why-use-testmain-for-testing-in-go-dafb52b406bc)

> Since function names need to be unique in a given package, you can only define the TestMain once for each package. If your package has multiple test files, choose a logic place for your single TestMain function.

File: models/keyword_test.go (hunk 7 above)
Comment (gutakk) [reply models/keyword_test.go]: As I used `ReadAll()` function and the signature look like this
```go
func (r *Reader) ReadAll() (records [][]string, err error) {...}
```

or is it better to loop through file and use `Read()` to construct slice manually instead?

File: models/keyword.go (hunk 4 above)
Comment (gutakk) [reply models/keyword.go]: We can discuss about this in this [comment](https://github.com/gutakk/go-google-scraper/pull/25#discussion_r532476260)

File: controllers/keyword.go (hunk 1 above)
Comment (gutakk) [reply controllers/keyword.go]: Fixed in 3dcfa7705031d36383aba55681a40f9a7c94ddc0

File: models/keyword.go (hunk 3 above)
Comment (gutakk) [reply models/keyword.go]: No it doesn't. Fixed in 56662ee0b4df157d4d83680122d71b54be3a45aa

File: models/keyword_test.go (hunk 5 above)
Comment (gutakk) [reply models/keyword_test.go]: Good catch. it is impossible. Fixed in b2b3d88f76e43e6cdf496f6033eac0fce9a58c05

File: models/keyword_test.go (hunk 6 above)
Comment (gutakk) [reply models/keyword_test.go]: Fixed in b02cbc6b0abe3720c279b654eaecdd704f614207

File: controllers/keyword.go (hunk 2 above)
Comment (gutakk) [reply controllers/keyword.go]: Fixed in 45e248a014c3d66aa52efa2092fa3bc3f97eb9c4

File: models/keyword_test.go (hunk 7 above)
Comment (gutakk) [reply models/keyword_test.go]: I think it is much more better to use `Read()` to create `[]string` slice
Fixed in cd31b5299e42b4d28439db3a244e1cf1d872adae

File: controllers/keyword.go (hunk 2 above)
Comment (olivierobert) [reply controllers/keyword.go]: There could also be another model for KeyworsdFile specifically for the file validation 💡

File: models/setup_test.go (hunk 8 above)
Comment (olivierobert) [reply models/setup_test.go:10]: Okay, so basic reason indeed 😅

File: controllers/keyword_test.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+	user := models.User{Email: email, Password: string(hashedPassword)}
+	db.GetDB().Create(&user)
+
+	formData := url.Values{}
+	formData.Set("email", email)
+	formData.Set("password", password)
+
+	headers := http.Header{}
+	headers.Set("Content-Type", "application/x-www-form-urlencoded")
+	response := testHttp.PerformRequest(s.engine, "POST", "/login", headers, formData)
```
Comment (olivierobert) [inline controllers/keyword_test.go]: Sorry for not noticing this before, but having to manually log in for a unit tests seem awkward and not standard 🤔 Wouldn't saving a cookie as a fixture and set it manually would not work? Let me know if this is because of a limitation of Gin.

File: helpers/error_handler/error_message.go [hunk 10]
```diff
@@ -44,6 +46,8 @@ func DatabaseErrorMessage(err error) error {
 	pgErr := err.(*pgconn.PgError)
 
 	switch pgErr.Code {
+	case foreignKeyErrorCode:
+		return errors.New(somethingWentWrongError)
```
Comment (olivierobert) [inline helpers/error_handler/error_message.go:50]: Having so generic message is not so nice and it could create debugging issues down the road 😵 I don't have a better option now

File: helpers/error_handler/error_message.go [hunk 11]
```diff
@@ -8,14 +8,16 @@ import (
...
-	minError            = "%s must be longer than %s"
-	passwordEqError     = "Passwords do not match"
-	requiredError       = "%s is required"
-
-	duplicateErrorCode = "23505"
+	commonFieldError        = "%s is not valid"
+	emailFormatError        = "Invalid email format"
+	emailDuplicateError     = "Email already exists"
+	minError                = "%s must be longer than %s"
+	passwordEqError         = "Passwords do not match"
```
Comment (olivierobert) [inline helpers/error_handler/error_message.go]: After some more Go conventions reading, errors should always be in lowercase: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: helpers/user/current_user.go [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/gutakk/go-google-scraper/models"
+)
+
+func GetCurrentUser(c *gin.Context) models.User {
+	currentUser := c.MustGet("currentUser")
+
+	if currentUser != nil {
+		return currentUser.(models.User)
```
Comment (olivierobert) [inline helpers/user/current_user.go]: I am not sure why using [a type assertion](https://golang.org/ref/spec#Type_assertions) here as per the usage in the controller:

```go
currentUser := helpers.GetCurrentUser(c)
```

With the type assertion, shouldn't this be:

```go
currentUser, ok := helpers.GetCurrentUser(c)
```

File: controllers/keyword_test.go [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+	engine := testConfig.GetRouter(true)
+	new(KeywordController).applyRoutes(EnsureAuthenticatedUserGroup(engine))
+
+	response := testHttp.PerformRequest(engine, "POST", "/keyword", nil, nil)
+
+	assert.Equal(t, http.StatusFound, response.Code)
+	assert.Equal(t, "/login", response.Header().Get("Location"))
+}
+
+func TestDisplayKeywordWithAuthenticatedUser(t *testing.T) {
```
Comment (olivierobert) [inline controllers/keyword_test.go]: Minor thing but shouldn't we group all tests cases for display and upload together. Now it seems messy which makes it hard to follow:

```
upload - TestUploadKeywordWithValidParams
upload - TestUploadKeywordWithBlankPayload
display - TestDisplayKeywordWithGuestUser
upload - TestUploadKeywordWithGuestUser
display - TestDisplayKeywordWithAuthenticatedUser
```

I am not sure now if `TestUploadKeywordWithValidParams` and `TestUploadKeywordWithBlankPayload` are for guest or authenticated users unless I check carefully the whole files.

It's more standard to group tests by:
- Order of appearance of the methods in the file under tests. Since the simply method appears first, then all tests for display should be first. Then the tests for upload can come after that.
- Happy path before sad path: so tests authenticated user should come before tests with no authenticated users within the group of tests for each method.

File: helpers/error_handler/error_message.go (hunk 10 above)
Comment (gutakk) [reply helpers/error_handler/error_message.go:50]: I also don't have better solution for this that's why I provided generic message ATM 😅

File: helpers/user/current_user.go (hunk 12 above)
Comment (gutakk) [reply helpers/user/current_user.go]: Because `MustGet("currentUser")` return interface type so I need to cast it to `models.User` type so I can use its attribute. As `currentUser` can be any type so I added the type check in 891f46f42d4ef5ab6165819c5744526a5bd28996

File: helpers/error_handler/error_message.go (hunk 11 above)
Comment (gutakk) [reply helpers/error_handler/error_message.go]: Fixed in 43464969e9fc3212187adb73332eb81732a3e059

File: controllers/keyword_test.go (hunk 13 above)
Comment (gutakk) [reply controllers/keyword_test.go]: Good catch Fixed in 423332f7a1b2b4fa1a7f52edd71809c496ddafaa.

Maybe should this conversation be a part of our Go convention?

File: controllers/keyword_test.go (hunk 9 above)
Comment (gutakk) [reply controllers/keyword_test.go]: Good catch. Finally I found the way to save a cookie as a fixture in 4f3d586812302dec2af69154f873a02ea3475a22

File: controllers/keyword_test.go (hunk 13 above)
Comment (olivierobert) [reply controllers/keyword_test.go]: This seems more like general best practice for all languages but if you feel it's needed to specify for Go then go for it.

File: helpers/user/current_user.go (hunk 12 above)
Comment (olivierobert) [reply helpers/user/current_user.go]: Now with the type check, it does make more sense 😁

File: helpers/str/str.go [hunk 14]
```diff
@@ -0,0 +1,13 @@
+package str
```
Comment (olivierobert) [inline helpers/str/str.go]: Instead of `str` `text` would be more descriptive 💭

File: helpers/html/render_html.go [hunk 15]
```diff
@@ -4,13 +4,14 @@ import (
...
+	"github.com/gutakk/go-google-scraper/helpers/str"
 )
 
 func RenderWithError(c *gin.Context, status int, view string, title string, err error, data map[string]interface{}) {
 	ginview.HTML(c, status, view, gin.H{
-		"title":  title,
-		"errors": err.Error(),
-		"data":   data,
+		"title": title,
+		"error": str.CapitalizeFirst(err.Error()),
```
Comment (olivierobert) [inline helpers/html/render_html.go]: Not sure if `CapitalizeFirst` is really needed. The capitalization is more of a UI concern and could be solved with CSS 😄 Say, it was an API then it might make sense but for a web app, showing humane and well-formatted errors can be solved in the UI.

File: controllers/keyword_test.go [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+	response := testHttp.PerformRequest(engine, "GET", "/keyword", nil, nil)
+
+	assert.Equal(t, http.StatusFound, response.Code)
+	assert.Equal(t, "/login", response.Header().Get("Location"))
+}
+
+func (s *KeywordDbTestSuite) TestUploadKeywordWithAuthenticatedUserAndValidParams() {
+	headers, payload := testFile.CreateMultipartPayload("tests/fixture/adword_keywords.csv")
+	cookie := fixture.GetCookie("user_id", s.userID)
```
Comment (olivierobert) [inline controllers/keyword_test.go]: I am not sure if it should be called `GetCookie` as it seems to actually generate encoded cookie data unless I am missing something in the helper code 🤔 So `fixture.GenerateCookie("user_id", s.userID)` might make more sense.

File: tests/csv/adword_keywords.csv [hunk 17]
```diff
@@ -0,0 +1 @@
+AWS
```
Comment (olivierobert) [inline tests/csv/adword_keywords.csv]: This is a duplicate of the file `fixture/adword_keywords.csv` so it can be removed

**Review by olivierobert — COMMENTED** [review]
Minor fixes to and then good to go 💪

File: controllers/keyword_test.go (hunk 16 above)
Comment (gutakk) [reply controllers/keyword_test.go]: Make more sense. Fixed in c2abe467ca7f77b75aeb578ec9b69285bc0026ab

File: tests/csv/adword_keywords.csv (hunk 17 above)
Comment (gutakk) [reply tests/csv/adword_keywords.csv]: 😮 My bad. Fixed in 4cc2fbcd1abd04500ccc952eec2ae4864c9c9bef

File: helpers/html/render_html.go (hunk 15 above)
Comment (gutakk) [reply helpers/html/render_html.go]: Fixed in 2190a389763d3c2b8f161a17240f653665c92f63
I can't find any resources but I assume that Go unit test doesn't include css stuff. Thus I also need to update the unit test.

File: helpers/str/str.go (hunk 14 above)
Comment (gutakk) [reply helpers/str/str.go]: According to this [comment](https://github.com/gutakk/go-google-scraper/pull/25#discussion_r536014123). I removed `str package`

File: helpers/html/render_html.go (hunk 15 above)
Comment (olivierobert) [reply helpers/html/render_html.go]: In a unit test, the only test for lowercase, it's fine. For UI test, we also test for [case insensitivity](https://til.hashrocket.com/posts/85bda9479b-expect-a-case-insensitive-match-) to avoid tests breaking when the UI changes.

**Review by olivierobert — APPROVED** [review]
